core/src/main/scala/com/salesforce/op/stages/impl/insights/RecordInsightsLOCO.scala

```diff
@@ -135,65 +162,96 @@ class RecordInsightsLOCO[T <: Model[T]]
+
+      // Let's check the indicator value and descriptor value
+      // If those values are empty, the field is likely to be a derived text feature (hashing tf output)
+      if (history.indicatorValue.isEmpty && history.descriptorValue.isEmpty) {
+        // Name of the field
+        val rawName = history.parentFeatureStages match {
+          case s if s.exists(_.contains(smartTextClassName)) => history.parentFeatureOrigins.headOption
+          case s if s.exists(_.contains(smartTextMapClassName)) => history.grouping
+          case _ => None
+        }
+        rawName.foreach { name =>
+          // Update the aggregation map
+          val (indices, array) = aggregationMap.getOrElse(name, (Array.empty[Int], Array.empty[Double]))
+          aggregationMap.update(name, (indices :+ i, sumArray(array, diffToExamine)))
+        }
+      } else addToHeaps(i, max, diffToExamine)
+    }
+
+    // Adding LOCO results from aggregation map into heaps
+    for {(indices, ar) <- aggregationMap.values} {
+      // The index here is arbitrary
+      val i = indices.head
+      val n = indices.length
+      val diffToExamine = ar.map(_ / n)
+      val max = diffToExamine(indexToExamine)
+      addToHeaps(i, max, diffToExamine)
+    }
+
     val topPositive = positiveMaxHeap.dequeueAll
     val topNegative = negativeMaxHeap.dequeueAll
-    (topPositive ++ topNegative)
+    topPositive ++ topNegative
   }
 
-  override def transformFn: OPVector => TextMap = (features) => {
+  override def transformFn: OPVector => TextMap = features => {
     val baseResult = modelApply(labelDummy, features)
     val baseScore = baseResult.score
 
     // TODO sparse implementation only works if changing values to zero - use dense vector to test effect of zeros
     val featuresSparse = features.value.toSparse
-    val featureArray = featuresSparse.indices.zip(featuresSparse.values)
-    val filledSize = featureArray.length
+    // Besides non 0 values, we want to check the text features as well
+    val indices = (featuresSparse.indices ++ textIndices).distinct.sorted
+    val featureArray = indices.map(i => i -> features.value(i))
     val featureSize = featuresSparse.size
 
     val k = $(topK)
+    // Index where to examine the difference in the prediction vector
     val indexToExamine = baseScore.length match {
       case 0 => throw new RuntimeException("model does not produce scores for insights")
       case 1 => 0
       case 2 => 1
-      case n if (n > 2) => baseResult.prediction.toInt
+      case n if n > 2 => baseResult.prediction.toInt
     }
-    val topPosNeg = returnTopPosNeg(filledSize, featureArray, featureSize, baseScore, k, indexToExamine)
+    val topPosNeg = returnTopPosNeg(featureArray, featureSize, baseScore, k, indexToExamine)
     val top = getTopKStrategy match {
       case TopKStrategy.Abs => topPosNeg.sortBy { case (_, v, _) => -math.abs(v) }.take(k)
       case TopKStrategy.PositiveNegative => topPosNeg.sortBy { case (_, v, _) => -v }
```

Comment:
why in case of `TopKStrategy.PositiveNegative` we dont apply `take(k)`?

Reply:
In `TopKStrategy.PositiveNegative`, we want to return `k` positives and `k` negatives.

Reply:
makes sense. so perhaps `take(2*k)`?